rust-runtime/aws-smithy-runtime-api/src/client/endpoints.rs

```diff
@@ -28,8 +34,77 @@ impl StaticUriEndpointResolver {
 }
 
 impl EndpointResolver for StaticUriEndpointResolver {
-    fn resolve_and_apply_endpoint(&self, request: &mut HttpRequest) -> Result<(), BoxError> {
+    fn resolve_and_apply_endpoint(
+        &self,
+        _params: &EndpointResolverParams,
+        _endpoint_prefix: Option<&EndpointPrefix>,
+        request: &mut HttpRequest,
+    ) -> Result<(), BoxError> {
         apply_endpoint(request.uri_mut(), &self.endpoint, None)?;
         Ok(())
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct DefaultEndpointResolver<Params> {
+    inner: SharedEndpointResolver<Params>,
+}
+
+impl<Params> DefaultEndpointResolver<Params> {
+    pub fn new(resolve_endpoint: SharedEndpointResolver<Params>) -> Self {
+        Self {
+            inner: resolve_endpoint,
+        }
+    }
+}
+
+impl<Params> EndpointResolver for DefaultEndpointResolver<Params>
+where
+    Params: Debug + Send + Sync + 'static,
+{
+    fn resolve_and_apply_endpoint(
+        &self,
+        params: &EndpointResolverParams,
+        endpoint_prefix: Option<&EndpointPrefix>,
+        request: &mut HttpRequest,
+    ) -> Result<(), BoxError> {
+        let endpoint = match params.get::<Params>() {
+            Some(params) => self.inner.resolve_endpoint(params)?,
+            None => {
+                return Err(Box::new(ResolveEndpointError::message(
+                    "params of expected type was not present",
+                )));
+            }
+        };
+
+        let uri: Uri = endpoint.url().parse().map_err(|err| {
+            ResolveEndpointError::from_source("endpoint did not have a valid uri", err)
+        })?;
+
+        apply_endpoint(request.uri_mut(), &uri, endpoint_prefix).map_err(|err| {
+            ResolveEndpointError::message(format!(
+                "failed to apply endpoint `{:?}` to request `{:?}`",
+                uri, request,
+            ))
+            .with_source(Some(err.into()))
+        })?;
+
+        for (header_name, header_values) in endpoint.headers() {
+            request.headers_mut().remove(header_name);
+            for value in header_values {
+                request.headers_mut().insert(
+                    HeaderName::from_str(header_name).map_err(|err| {
+                        ResolveEndpointError::message("invalid header name")
+                            .with_source(Some(err.into()))
+                    })?,
+                    HeaderValue::from_str(value).map_err(|err| {
+                        ResolveEndpointError::message("invalid header value")
+                            .with_source(Some(err.into()))
+                    })?,
```

Comment:
Should these errors even be possible if the endpoint config is valid?

Reply:
I am not entirely sure. I've decided to preserve the existing error handling logic from the [original PR](https://github.com/awslabs/smithy-rs/pull/2063) for the sake of porting. I don't see how an endpoint resolver comes into being after a series of config steps in the orchestrator world is all that different from the old world, so maybe it makes sense to keep the existing error handling for now (we can later turn them into non-recoverable errors, if we want).